cpp/src/arrow/io/CMakeLists.txt

```diff
@@ -20,6 +20,7 @@
 
 set(ARROW_IO_LINK_LIBS
   arrow_shared
+  dl
```

Comment:
This should not be needed, maybe you mixed somewhere a static and a shared build?